Samples/NetworkMediaStreamer/RTPMediaStreamer/src/RTPStreamSink.cpp

```diff
@@ -0,0 +1,433 @@
+                pt[2] = sc[3];
+                p.Length(3);
+                profileIdc = winrt::to_string(winrt::Windows::Security::Cryptography::CryptographicBuffer::EncodeToHexString(p));
+            }
+            memcpy_s(p.data(), p.Capacity(), sc, nalsz);
+            p.Length((uint32_t)nalsz);
+            auto paramSet = winrt::Windows::Security::Cryptography::CryptographicBuffer::EncodeToBase64String(p);
+            paramSets += winrt::to_string(paramSet) + ",";
+            sc = sc1;
+        }
+    }
+    if (!paramSets.empty())
+    {
+        //remove last comma
+        paramSets.pop_back();
+    }
+    auto sep = destination.find(":");
+    auto destIP = destination.substr(0, sep);
+    auto destPort = destination.substr(sep + 1, destination.find("?") - sep);
+    std::string sdp =
+        "v=0\n"
+        "o=- " + std::to_string(rand()) + " 0 IN IP4 127.0.0.1\n" //destination
+        "s=MSFT VideoStreamer\n"
+        "c=IN IP4 " + destIP + "\n" //source
+        "t=0 0\n"
+        "m=video " + destPort + " RTP/AVP " + std::to_string(h264payloadType) + "\n"
+        "a=ts-refclk:ntp=time.windows.com\n"
+        "a=rtpmap:" + std::to_string(h264payloadType) + " H264/90000\n"
+        "a=fmtp:" + std::to_string(h264payloadType) + " packetization-mode=" + std::to_string(m_packetizationMode);
+    if (!paramSets.empty())
+    {
+        sdp = sdp
+            + "; sprop-parameter-sets="
+            + paramSets + "; profile-level-id=" + profileIdc + "\n";
+    }
+    memcpy_s(buf, maxSize, sdp.c_str(), sdp.size());
+    buf[sdp.size()] = 0;
+    HRESULT_EXCEPTION_BOUNDARY_END;
+}
+
+//IMFSampleGrabberCallback
+STDMETHODIMP RTPVideoStreamSink::PacketizeAndSend(IMFSample* pSample)
+{
+    auto lock = std::lock_guard(m_guardlock);
+    winrt::com_ptr<IMFMediaBuffer> spMediaBuf;
+    LONGLONG llSampleTime, llSampleDur;
+    DWORD dwSampleSize, maxLen;
+    BYTE* pSampleBuffer = nullptr;
+    HRESULT hr = S_OK;
+    try {
+
+        winrt::check_hresult(pSample->GetBufferByIndex(0, spMediaBuf.put()));
+        winrt::check_hresult(spMediaBuf->Lock(&pSampleBuffer, &maxLen, &dwSampleSize));
+        winrt::check_hresult(pSample->GetSampleTime(&llSampleTime));
+        winrt::check_hresult(pSample->GetSampleDuration(&llSampleDur));
+        MFTIME latency = MFGetSystemTime() - llSampleTime;
+        //std::cout << "\nlatency = " << latency / 10000;
+        auto ts = ((llSampleTime) * 90) / 10000;
+        if (m_packetizationMode == 1)
+        {
+            PacketizeMode1(pSampleBuffer, dwSampleSize, ts);
+        }
+        else
+        {
+            PacketizeMode0(pSampleBuffer, dwSampleSize, ts);
+        }
+    }
+    catch (winrt::hresult_error const& ex)
+    {
+        hr = ex.code();
+    }
+
+    if (spMediaBuf && pSampleBuffer)
+    {
+        spMediaBuf->Unlock();
+    }
+    return hr;
+}
+
+INetworkMediaStreamSink* RTPVideoStreamSink::CreateInstance(IMFMediaType* pMediaType, IMFMediaSink* pParent, DWORD dwStreamID)

```

Comment:
should this code throw

Reply:
that is ok, this is called within the same dll and then caught at the exception boundary of CreateRTPMediaSink

---
In reply to: [491170636](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r491170636) [](ancestors = 491170636)